group/views.py

```diff
@@ -102,41 +95,58 @@ def get_all_announce(request, group_id):
 
 def detail(request, group_id):
 
+    redirect_id = 1; #return to detail page.
     group = get_group(group_id)
     show_number = 5; #number for brief list to show in group detail page.
-    all_contest = group.trace_contest.all()
+    all_contest = group.trace_contest.order_by('-start_time')
     annowence_list = group.announce.all()
     student_list = group.member.order_by('user_level')
-    coowner_list = group.coowner.all()
     owner = group.owner
-    user_is_owner = has_group_ownership(request.user, group)
+    form = AnnounceForm()
+
+    user = validate_user(request.user)
+    user_is_owner = has_group_ownership(user, group)
+    user_is_coowner = has_group_coownership(user, group)
+
+    user_has_auth = user_is_owner or user_is_coowner
 
     running_contest_list = []
     ended_contest_list = []
     now = timezone.now()
-    for contest in all_contest:
-        if contest.start_time < now and contest.end_time > now:
-            running_contest_list.append(contest)
-        elif contest.end_time < now:
-            ended_contest_list.append(contest)
+    running_contest_list = group.trace_contest.filter(start_time__lte=now, end_time__gte=now)
+    ended_contest_list = group.trace_contest.filter(end_time__lte=now)
+
+    student_list = get_current_page(request, student_list)
 
     return render_index(
         request, 'group/groupDetail.html', {
-            'rc_list': running_contest_list[0:show_number],
-            'ec_list': ended_contest_list[0:show_number],
-            'an_list': annowence_list,
-            'coowner_list': coowner_list,
+            'running_contest_list': running_contest_list[0:show_number],
+            'ended_contest_list': ended_contest_list[0:show_number],
+            'announce_list': annowence_list,
             'owner': owner,
-            's_list': student_list,
-            'group_name': group.gname,
+            'student_list': student_list,
+            'group_name': group.gname, 
```

Comment:
Is it possible to pass a group object to your template, then use ```{{ group.name }}``` 
(instead of ```{{ group_name }}```)to access group name.
To do so, you can pass fewer variable.